source/common/init/target_impl.cc

```diff
@@ -27,6 +27,10 @@ TargetImpl::TargetImpl(absl::string_view name, InitializeFn fn)
       fn_(std::make_shared<InternalInitalizeFn>([this, fn](WatcherHandlePtr watcher_handle) {
         watcher_handle_ = std::move(watcher_handle);
         fn();
+        // It is possible that fn() mutates the is_ready_.
```

Comment:
I don't think this logic is correct. The only way that the initialization `fn()` can mutate `is_ready_` is by calling `ready()`. You don't want to call it twice.

Reply:
The fact is that `TargetImpl::ready()` is idempotent. I see several places utilize this assumption.

Is the lambda abusing the fact? If so I can fix

Reply:
True, it's idempotent, but that's for the user's benefit. To me, it feels really weird to say:
```c++
if (is_ready_) {
  ready();
}
```
It also feels wrong to call `fn()`, and unnecessary to hang onto the watcher handle, if `is_ready_` is true.

Reply:
fixed in the latest commit